### ISSUE TITLE
Event logs for maintenance and hibernation operations

### ISSUE THREAD

**Issue by danielfoehrKn** [gardener/gardener#2390]
**How to categorize this issue?**

/area operations
/area usability
/kind enhancement
/priority normal

**What would you like to be added**:

Add an "event-log" of performed maintenance operations as well as hibernation / wake-up operations.

Below I listed three options that I could think of how to realize this. 

Option (1)

Add an events (or otherwise named field) into the Shoot.Status containing arbitrary Shoot events identified by the type field.

``` yaml
status:
  conditions: <>
  hibernated: true
  events:
    - type: MachineImageVersionMaintenance
      timestamp: "2020-05-25T14:47:54Z"
      status: "Successful"
      message: "Updated image of worker-pool 'coreos-1' from 'coreos' version 'xy' to version 'abc'."
      reason: "AutoUpdate of MachineImage configured"
    - type: MachineImageVersionMaintenance
      timestamp: "2020-05-25T14:47:53Z"
      status: "Successful"
      message: "Updated image of worker-pool 'coreos-2' from 'coreos' version 'xy' to version 'abc'."
      reason: "Machine image version is expired"
    - type: KubernetesVersionMaintenance
      timestamp: "2020-05-25T14:47:52Z"
      status: "Successful"
      message: "Updated Kubernetes version '0.0.1' to version '0.0.5'. This is an increase in the patch level."
      reason: "AutoUpdate of Kubernetes version configured"
    - type: KubernetesVersionMaintenance
      timestamp: "2020-05-25T14:47:53Z"
      status: "Successful"
    - type: KubernetesVersionMaintenance
      timestamp: "2020-05-24T14:47:53Z"
      status: "Successful"
    - type: Hibernation
      timestamp: "2020-05-24T14:47:53Z"
    - type: WakeUp
      timestamp: "2020-05-26T14:47:53Z"
```

Option (2)
To have a separation between Maintenance and Hibernation, this would be another option.
This would also group events that happened during the same maintenace time window.
``` yaml
status:
  conditions: <>
  hibernated: true
  maintenance:
    - lastSuccessfulMaintenance: "2020-05-25T14:47:53Z"
      events:
      - timestamp: "2020-05-25T14:47:53Z"
        status: "Successful"
        messages:
          - message: Updated image of worker-pool 'coreos-1' from 'coreos' version 'xy' to version 'abc'.
            reason: AutoUpdate of MachineImage configured
            type: MachineImageVersion
          - message: Updated Kubernetes version '0.0.1' to version '0.0.5'. This is an increase in the patch level.
            reason: AutoUpdate of Kubernetes version configured.
            type: KubernetesVersion
  hibernation:
    - timestamp: "2020-05-25T14:47:53Z"
      type: "Hibernation"
    - timestamp: "2020-05-26T14:47:53Z"
      type: "WakeUp"
```

Another option (3) would be to have a dedicated resource in the core.gardener-cloud API group 

``` yaml
apiVersion: core.gardener.cloud/v1beta1
kind: ShootOperation
metadata:
  name: <shoot-name>
  namespace: shoot--dev--<>
  ownerReferences:
    - kind: Shoot
      name: <shoot-name>
spec:
  type: Maintenance / Hibernation
status:
  lastOperation: "2020-05-21T00:00:12Z"
  events: # only contain actual changes
    - type: MachineImageVersionMaintenance
      timestamp: "2020-05-25T14:47:54Z"
      status: "Successful"
      message: "Updated image of worker-pool 'coreos-1' from 'coreos' version 'xy' to version 'abc'."
      reason: "AutoUpdate of MachineImage configured"
    - type: MachineImageVersionMaintenance
      timestamp: "2020-05-25T14:47:53Z"
      status: "Successful"
      message: "Updated image of worker-pool 'coreos-2' from 'coreos' version 'xy' to version 'abc'."
      reason: "Machine image version is expired"
    - type: KubernetesVersionMaintenance
      timestamp: "2020-05-25T14:47:52Z"
      status: "Successful"
      message: "Updated Kubernetes version '0.0.1' to version '0.0.5'. This is an increase in the patch level."
      reason: "AutoUpdate of Kubernetes version configured"
```

I tend to favour option (3) to not pollute the already quite large Shoot resource with Hibernation and Maintenance event logs. @vpnachev also suggested to restrict the TTL of an event to 14 days.
The same TTL could also be done if we store the event log in the external resource.

Another issue that I see is how to deal with failed Shoot operations e.g a Shoot using an expired Kubernetes version cannot be updated because of a `CloudProfile` misconfiguration.
It would be handy if the controller could persist this information without writing a new event for each re-tried operation. There are several ways to achieve this.

**Why is this needed**:

There is no place to properly store a log of automatically performed maintenance operations.
For instance for Maintenance, the Gardener Controller manager just writes events on the Shoot resource stating the reason for a performed update operation.
The problem is that Kubernetes events are Garbage collected after a certain amout of time ([configurable in the Kube-Apiserver](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/). 
Usually this is somewhere between 1 - 24 hours.

This is problematic because
- Users want to be able to see which maintenance operation Gardener automatically performed, when, why and if it was successful. 
- In case of unwanted update behaviour / problems, operators cannot easily determine what was the reason for the update operation.

In addition Gardener does currently only persist the current and desired state of the Hibernation.
Shoot owners might also want to see a log of the manually & automatically triggered Hibernations.

Integrators like the gardener Dashboard could read this information and display an operations log for both the Hibernation and the Maintenance operations.

**Comment by rfranzke** [comment]
cc @gardener/dashboard-maintainers

**Comment by BeckerMax** [comment]
Hey, I am new to K8s and Gardener so I very much welcome corrections/feedback to my suggestion below:

I think it would be cleaner to not introduce any new custom resource for retaining events. Instead, the already created events would best be recorded in some way and then saved in a storage form made for logs. In our case that would mean that something reads these events and writes them to the Elasticsearch(/Loki) instance accessible by the Shootowner. The user   can access them via  Kibana(/Grafana). There are multiple solutions like [kubernetes-event-exporter](https://github.com/opsgenie/kubernetes-event-exporter) or [event-router](https://github.com/heptiolabs/eventrouter) but I am worried they could not work in our setup: If we run these event exporters in the garden cluster can they actually contact the kibana instances in the seed clusters? What do you think? Do you have other ideas?
Advantages of this approach:
1. K8s docu itself recommends it [here](https://kubernetes.io/docs/tasks/debug-application-cluster/events-stackdriver/): `To provide longer history and aggregation capabilities, a third party solution should be installed to capture events.`. 
2. etcd is a key-value store not made for storing logs. There were problems in the past with high load on etcd due to events [here](https://github.com/kubernetes/kubernetes/issues/47532) and [here](https://github.com/kubernetes/kubernetes/issues/4432). This might not be very critical as we use a second etcd for events.
3. This might be a minor but by retaining events of customers in our etcd we are paying for storage which the customer should pay. If tens of thousands of clusters get createad/hibernated in two weeks this could add up.
4. We do not introduce new code (hopefully).

As an alternative we could implement a new resource as originally suggested. After discussion with @tim-ebert  we went away from using a resource which appends new events (as discussed originally above). Instead we want to create something similiar to the current kubernetes events but with a longer _ttl_ (time-to-live in etcd). We thought that would be closer to the original kubernetes implementation and therefore help understanding. Steps involved for the implementation:

- Create a new resource in our custom api-server. It should be similiar to the k8s event object. When registering the RESTStorageProvider set a 14 day ttl similiar to [here](https://github.com/kubernetes/kubernetes/blob/9fb34eddedca31c9e36b0adfeda6219a0efa6cb4/pkg/master/master.go#L437). Probably we can reuse the [storageProvider](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/events/rest/storage_events.go) code ?
-  Create a new eventRecorder. Currently, events are written via client-go [eventRecorder ](https://github.com/kubernetes/kubernetes/blob/9fb34eddedca31c9e36b0adfeda6219a0efa6cb4/staging/src/k8s.io/client-go/tools/events/event_recorder.go#L44) . It would be convenient to have a similar mechanism implementing the same [interface](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/events/interfaces.go) which then writes the events in the form of the new resource.
- Adapt the machine controller to use the new eventRecorder in addition for certain events.

WDYT?

**Comment by rfranzke** [comment]
Thanks @BeckerMax for thinking it through so thoroughly.

I think the approach of leveraging the existing `corev1.Event` resource and exporting it to a place where the data can be stored for a longer time makes sense in general.
However, I think this is something that needs to taken care of during the deployment of Gardener. The contract from Gardener's point of view could simply be to generate these events. If somebody wants to store them for a longer time then it's his responsibility to deploy additional tools into the garden cluster (e.g., the event-exporter + elasticsearch) and configure them properly to pull the events out of etcd. Consequently, if we would follow this approach, we wouldn't do anything about it here in the gardener/gardener repository other than generating these `corev1.Event` resources (and maybe providing a document explaining these 'best practices').

The disadvantage, however, would be that our ops tools (e.g., `gardenctl` or the dashboard) cannot "easily" display these events because they are no longer Kubernetes resources. They would be stored in third-party tools for which "special"/dedicated code needs to be developed. Similarly, end-users that work with the Gardener API programmatically would need to do the same. OTOH, they are already required to do it for accessing their controlplanes' monitoring or logging data.

Considering that in large landscapes most shoots won't have a hibernation schedule and won't get too often Kubernetes/machine image version updates I'd assume that we wouldn't have "too many" of such `ShootEvent` resources. Though, it can theoretically happen (based on the configuration of Gardener and the shoots, basically), and then we might most likely face bottlenecks and other issues as etcd is not made for storing such information.

Probably it would make sense to have a short discussion with @gardener/gardener-maintainers @gardener/dashboard-maintainers @vlvasilev @dguendisch @vlerenc about the subject matter, and then decide jointly about the desired way to proceed.

**Comment by dguendisch** [comment]
I think both arguments are valid: @BeckerMax with not misusing etcd for kind-of log storage but instead re-using our logging solution (by using one of the many events-to-logs solutions) and @rfranzke with a) events are low-volume as maintenance operations are pretty heavy-weight, so customers cannot induce a large amount of them and b) Dashboard or other ops tools would need to code against the logging stack instead of k8s to retrieve those logs.

So it probably boils down to the question where a Gardener operator or consumer wants to read those maintenance events and if this feature is important enough to justify new UIs and extensions in e.g. gardenctl.
Sure, a dedicated Dashboard UI would be great but so do many other features that "would be great" :)

For me personally, it would totally suffice to know that we direct events into our logging stack and e.g. provide a grafana (loki) dashboard, i.e. "Shoot maintenance history" that would directly filter up those events. That grafana dashboard could be part of the users grafana so that consumers can view this too.
This would avoid additional dashboard/gardenctl development and could have the additional benefit of having events (not only for shoot maintenance but generally) available in our logging solution.

**Comment by timebertt** [comment]
I agree with @dguendisch that both options make total sense.

However, I'm not sure, how such a logging implementation would look like for the first step, as we don't have any logging stack visible to the enduser in the garden cluster as of now. We only have a logging stack with the Shoot Control Plane in the Seed clusters (which is also brought down during hibernation). So I guess, we would need a new instance of kibana/grafana (at first just) for displaying these Shoot operation events. 

I also started thinking about, how a possible Dashboard integration of such events would then look like (if we would want to do it). So probably we would start with a simple section on the shoot details page displaying the latest events in chronological order. But at the next step, the Shoot owners might ask for an aggregated view of all operations on shoots in their project(s) and then the Gardener operator might want to have an aggregated view of all operation of a whole gardener system. Then we would want to filter, and so on and so on.... So to make this feature more and more useful, we will basically implement a lot of stuff, we could probably get out-of-the-box with already existing logging solutions in combination with an event-to-logs solution.

Long story short: im kind of split on the options and would indeed enjoy a discussion on the matter as @rfranzke proposed.

**Comment by vlerenc** [comment]
FYI/from former times, there is `Gardener Cluster Audit Logs` #236. Quite long ago, 2-3y, we discussed how we could preserve some activity log and went for the "cheap" events-based approach.

In fact, the original goal was to get an overview of all operations and their outcome and find the logs for those operations, e.g. by ID. Much like https://bosh.io 's https://bosh.io/docs/cli-v2/#tasks, which inspired that thought since it was always so useful.

However, there were issues with data volume, global TTL, etc. so we do not have that actually in place to the degree we want it.

Furthermore, to this day, not all operations make their appearance in the shoot resource, e.g. autonomous operations that happen in the seed like for instance when CA/MCM adds/removes/replaces a VM, which is visible from outside/by the end user and may have relevant impact to some workload. In these cases, the end user would like to know what happened and why (e.g. node became unhealthy / file system turned read-only and the VM was therefore replaced after the grace period or CA scaled a node down, because it was poorly utilised).

Let's address some of the points:
- @danielfoehrKn I would not extend/grow our (shoot) resources, so if we don't use events, it should be a separate resource or another sink.
- @BeckerMax You are certainly right, that we should not misuse events for log aggregation. However, I was more thinking about metadata in the events and actual logs elsewhere, similar to `bosh tasks`. Also, https://kubernetes.io/docs/tasks/debug-application-cluster/events-stackdriver is not representative for two reasons:
  - This recommendation is sensible as it applies to master VMs with local disks - which we don't have. We have pods with PVs.
  - It's a shameless plug to run on GCP and use Stackdriver. I generally don't find that OK in a global community project, but here we go.
- @BeckerMax Data volume is indeed of consideration, but this is by far not the only issue. What's with the teams that request a new shoot cluster for every change they do and almost (D)DoSed us in the beginning until we implemented near-infinite scalability by means of the gardenlets? What with the "space" for `secrets` and `configmaps` we provide in the virtual garden cluster. As long as these topics are not resolved, I am not so much concerned about events, but all of them should be solved eventually, yes, i.e. API throttling per project, https://kubernetes.io/docs/concepts/policy/resource-quotas in the virtual garden cluster, and events and their logs. As for the latter, my pragmatic proposal would be to go for the same approach as with the cluster logging/monitoring: define max TTL and max volume and go by whatever is smaller, e.g. try to keep the events/logs/metrics for 30d, but if a certain volume is reached, cut off the oldest entries.
- @rfranzke I also believe that we cannot be prescriptive here or assume a specific logging stack. As usual, we need to walk the line between Gardener as an Open Source project and what everybody does productively. The line between giving the configuration freedom, but having something out of the box that actually works. If someone needs to attach his company logging stack like Stackdriver, Splunk, or whatever, then this should be possible, but it would be important to have a default solution in place. In that sense, a decoupled/loosely-coupled  solution would be also my preferred choice.
- @rfranzke I also tend, as written above, to have the metadata as events and the actual operation logs shipped somewhere else. However, I am no longer sure, but haven't we had an operations ID 2y ago and could find the logs? I don't see that anymore in GCM or gardenlet, but that would be useful/required then.
- @rfranzke As for retrieving the actual logs, as you have remarked already, `gardenctl` already has to interact with logging stacks and that will require a rewrite anyways soon (cc @DockToFuture) when we switch to Loki. Make it pluggable and you have a nice solution.
- @dguendisch and @tim-ebert Same is true for the dashboard, so if we fetch only the metadata from the events, fetching the actual logs should be pluggable (into `gardenctl` and `dashboard`) and we may deliver the first implementation for Loki.

WDYT? A discussion together would be more than welcome. :-)

**Comment by BeckerMax** [comment]
Outcome of the joined discussion:

As an immediate solution we want to extend the life-span of certain shoot related events in etcd. We start with events recording  'K8s version updates', 'imageversion updates' and hibernation. This is implemented with an event controller in PR https://github.com/gardener/gardener/pull/2649 .

To improve observability we need to further discuss :
-> Which relevant information for the users/operators are we currently missing?
-> Is the missing information best provided in logs or events?
-> How do we want to persist events? (Probably we want to export events to some persistent storage other than etcd).
-> Through which interfaces do we want to make the events accessible to the user?
-> Is it helpful for the user to have an `operations-id` to match events with logs?

Please feel free to add information to the above.

**Comment by rfranzke** [comment]
> -> Which relevant information for the users/operators are we currently missing?
> -> Is the missing information best provided in logs or events?

Do we already have events for reconciliation/deletion operations? 

> -> How do we want to persist events? (Probably we want to export events to some persistent storage other than etcd).
> -> Through which interfaces do we want to make the events accessible to the user?

Hm, didn't we decide on the event controller approach for now with a separate etcd for the events (which is a best practice anyways) and a "short/normal" TTL for the non-shoot events? Everything else is much more work and we don't have a clear plan now, so I'd suggest to postpone this discussion until we have gained some experience with the event controller approach.

> -> Is it helpful for the user to have an operations-id to match events with logs?

Probably yes - we had it earlier (2 years ago), but then kicked it out again because we had no solution to store the events for shoots for longer time. Now, with the event controller, we can introduce the operations ID again to allow to find/correlate logs easier.
